src/marks/auto.js

```diff
@@ -0,0 +1,328 @@
+import {barX, barY} from "./bar.js";
+import {rect, rectX, rectY} from "./rect.js";
+import {cell} from "./cell.js";
+import {frame} from "./frame.js";
+import {bin, binX, binY} from "../transforms/bin.js";
+import {group, groupX, groupY} from "../transforms/group.js";
+import {marks} from "../plot.js";
+import {ascending} from "d3";
+
+// https://github.com/observablehq/stdlib/blob/main/src/table.js
+const nChecks = 20; // number of values to check in each array
+
+export function auto(data, {x, y, fx, fy, color, size, mark} = {}) {
+  // Shorthand: array of primitives should result in a histogram
+  if (x === undefined && y === undefined && arrayIsPrimitive(data)) x = (d) => d;
+
+  // Allow x and y and other dimensions to be specified as shorthand
+  // field names (but note that they can also be specified as a
+  // {transform} object such as Plot.identity).
+  if (!isOptions(x)) x = makeOptions(x);
+  if (!isOptions(y)) y = makeOptions(y);
+  if (!isOptions(fx)) fx = makeOptions(fx);
+  if (!isOptions(fy)) fy = makeOptions(fy);
+  if (!isOptions(color)) color = makeOptions(color);
+  if (!isOptions(size)) size = makeOptions(size);
+
+  const {value: xValue} = x;
+  const {value: yValue} = y;
+
+  const {value: sizeValue, reduce: sizeReduce} = size;
+
+  // Determine the default reducer, if any.
+  let {reduce: xReduce} = x;
+  let {reduce: yReduce} = y;
+  if (xReduce === undefined)
+    xReduce = yReduce == null && xValue == null && sizeValue == null && yValue != null ? "count" : null;
+  if (yReduce === undefined)
+    yReduce = xReduce == null && yValue == null && sizeValue == null && xValue != null ? "count" : null;
+
+  const {zero: xZero} = x;
+  const {zero: yZero} = y;
+  const {value: colorValue, reduce: colorReduce} = color;
+  const {value: fxValue} = fx;
+  const {value: fyValue} = fy;
+
+  // TODO Default sizeReduce for ordinal/ordinal case?
+  // TODO Allow x: {field: "red"}, too?
+  // TODO We might need maybeColorChannel to detect constant colors?
+  // TODO The line mark will need z?
+  // TODO Limit and sort for bar charts (e.g. alphabet)?
+  // TODO Support array of primitives with no channels?
+  // TODO Look at Plot warnings and see how many we can prevent
+  // TODO Default to something other than turbo for continuous? Like:
+  //      scheme: (colorValue && isContinuous(color)) || colorReduce ? "ylgnbu" : undefined
+
+  // To apply heuristics based on the data types (values), realize the
+  // columns. We could maybe look at the data.schema here, but Plot’s
+  // behavior depends on the actual values anyway, so this probably is
+  // what we want. By materializing the columns here, we also ensure
+  // that they aren’t re-computed later in Plot.plot.
+  x = valueof(data, xValue);
+  y = valueof(data, yValue);
+  color = valueof(data, colorValue);
+  size = valueof(data, sizeValue);
+  fx = valueof(data, fxValue); // TODO Should we still materialize if heuristic doesn't depend on it?
+  fy = valueof(data, fyValue);
+
```

Comment:
Something like this for the default sizeReduce in the ordinal-ordinal case?

```js
if (sizeReduce === undefined && sizeValue == null && xReduce == null && yReduce == null && isOrdinal(x) && isOrdinal(y)) {
  sizeReduce = "count";
}
```

I dunno if you also want to check that `mark === undefined || mark === "dot"`… maybe?